simulator/src/main/java/com/hedera/block/simulator/grpc/impl/ConsumerStreamGrpcClientImpl.java

```diff
@@ -0,0 +1,121 @@
+import edu.umd.cs.findbugs.annotations.NonNull;
+import io.grpc.ManagedChannel;
+import io.grpc.ManagedChannelBuilder;
+import io.grpc.stub.StreamObserver;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import javax.inject.Inject;
+
+/**
+ * Implementation of {@link ConsumerStreamGrpcClient} that handles the consumption of blocks
+ * via gRPC streaming. This implementation manages the connection to the server and tracks
+ * metrics related to block consumption.
+ */
+public class ConsumerStreamGrpcClientImpl implements ConsumerStreamGrpcClient {
+    // Configuration
+    private final GrpcConfig grpcConfig;
+
+    // Service dependencies
+    private final MetricsService metricsService;
+
+    // gRPC components
+    private ManagedChannel channel;
+    private BlockStreamServiceGrpc.BlockStreamServiceStub stub;
+    private StreamObserver<SubscribeStreamResponse> consumerStreamObserver;
+
+    // State
+    private final List<String> lastKnownStatuses;
+
+    /**
+     * Constructs a new ConsumerStreamGrpcClientImpl with the specified configuration and metrics service.
+     *
+     * @param grpcConfig The configuration for gRPC connection settings
+     * @param metricsService The service for recording consumption metrics
+     * @throws NullPointerException if any parameter is null
+     */
+    @Inject
+    public ConsumerStreamGrpcClientImpl(
+            @NonNull final GrpcConfig grpcConfig, @NonNull final MetricsService metricsService) {
+        this.grpcConfig = requireNonNull(grpcConfig);
+        this.metricsService = requireNonNull(metricsService);
+        this.lastKnownStatuses = new ArrayList<>();
+    }
+
+    @Override
+    public void init() {
+        channel = ManagedChannelBuilder.forAddress(grpcConfig.serverAddress(), grpcConfig.port())
+                .usePlaintext()
+                .build();
+        stub = BlockStreamServiceGrpc.newStub(channel);
+        lastKnownStatuses.clear();
+    }
+
+    @Override
+    public void requestBlocks(long startBlock, long endBlock) throws InterruptedException {
+        Preconditions.requireWhole(startBlock);
+        Preconditions.requireWhole(endBlock);
+
+        CountDownLatch streamLatch = new CountDownLatch(1);
+        consumerStreamObserver = new ConsumerStreamObserver(metricsService, streamLatch, lastKnownStatuses);
+
+        SubscribeStreamRequest request = SubscribeStreamRequest.newBuilder()
+                .setStartBlockNumber(startBlock)
+                .setEndBlockNumber(endBlock)
+                .setAllowUnverified(true)
+                .build();
+        stub.subscribeBlockStream(request, consumerStreamObserver);
+
+        streamLatch.await();
+    }
+
+    @Override
+    public void completeStreaming() throws InterruptedException {
+        consumerStreamObserver.onCompleted();
+        // todo(352) Find a suitable solution for removing the sleep
+        Thread.sleep(100);
```

Comment:
I'm not 100% sure but I think Netty will continue to send BlockItems to onNext() as long as the channel is open. Right now, I think `onCompleted()` will only release the starting thread waiting at line 97. You might need to close the channel in this method and then release the starting thread.

Reply:
Hm, you're right. Fixed it.